src/sparseml/modifiers/smoothquant/pytorch.py

```diff
@@ -0,0 +1,217 @@
+
+
+class SmoothQuantModifierPyTorch(SmoothQuantModifier):
+    """
+    PyTorch implementation of the SmoothQuant algorithm
+
+    :param calibration_function: optional function to use for the forward pass, or None
+    to use the default tensor_module_forward
+    """
+
+    calibration_function: Optional[Callable] = None
+    hooks_: List = None
+
+    def on_initialize(self, state: State, **kwargs) -> bool:
+        """
+        Initialize and run SmoothQuant on the given state
+
+        :param state: state to run SmoothQuant on
+        :return: True on a successful run, False otherwise
+        """
+        super(SmoothQuantModifierPyTorch, self).on_initialize(state, **kwargs)
+
+        calibration_dataloader = state.data.calib
+        self.hooks_ = []
+
+        self._setup_scale_hooks()
+        self._calibrate(state.model, calibration_dataloader)
+        self._apply_smoothing()
+
+        return True
+
+    def on_finalize(self, state: State, **kwargs) -> bool:
+        """
+        Clean up by clearing the CUDA cache
+
+        :param state: unused
+        :return: True
+        """
+        super(SmoothQuantModifierPyTorch, self).on_finalize(state, **kwargs)
+        torch.cuda.empty_cache()
+
+        return True
+
+    @torch.no_grad()
+    def _resolve_mappings(self, model: ModifiableModelPyTorch):
+        """
+        Transforms the list of activations to smooth and their corresponding weights
+        into SmoothQuantMapping objects, resolving regular expressions.
+
+        For each activation in the mapping list, we find the corresponding weight to
+        balance by searching for the longest substring. For instance, if our balance
+        weight is ".*re:.*q_proj" and the activation is "re:.*self_attn_layer_norm" we
+        would match model.layer.0.p_proj to model.layer.0.self_attn_layer_norm and
+        repeat for model.layer.1 and so on
+        """
+        resolved_mappings = []
+        for to_balance, to_smooth in self.mappings:
+            to_smooth_layers = model.get_layers(to_smooth)
+            for layer_name, smooth_layer in to_smooth_layers.items():
+                if layer_name not in self.ignore:
+                    balance_layers = []
+                    for balance_suffix in to_balance:
+                        # find the submodule that matches the activation layer
+                        _, balance_layer = get_matching_layer(
+                            balance_suffix, layer_name, model.model
+                        )
+                        if balance_layer:
+                            balance_layers.append(balance_layer)
+                    # each mapping can contain multiple layers to balance, but only
+                    # one layer to smooth
+                    mapping = SmoothQuantMapping(
+                        layer_name, smooth_layer, balance_layers
+                    )
+                    resolved_mappings.append(mapping)
+        return resolved_mappings
```

Comment:
Is this the same as the one in base class but wrapped in torch.no_grad? if that's the case we could avoid duplicacy by doing something like `func_to_wrap = wrapper(func_to_wrap)`?

Reply:
Ah good catch, this function doesn't actually need torch.no_grad() to run, I fully ported it over to base.py and forgot to delete to pytorch one